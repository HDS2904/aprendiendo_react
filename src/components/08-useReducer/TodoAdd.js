import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ( { handleAdd } ) => {
	const [ { description }, handleInputChange, reset ] = useForm({ description: ''});

	const handleSubmit = (e) => {
		e.preventDefault();

		if(e.description?.trim().length <= 0){
			return;
		}

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false
		}

		const action = {
			type: 'add',
			payload: newTodo
		}

		handleAdd(action)
		reset();
	}

	return (
		<>
			<h4>Agregar TODO</h4>
			<hr />
				<form onSubmit={handleSubmit}>
					<input 
						className='form-control'
						type="text" 
						name="description"
						autoComplete='off'
						placeholder='Actividad...'
						value={description}
						onChange={handleInputChange}
					/>
					<button
						className='btn btn-outline-primary mt-3 btn-block'
						type='submit'
					>
						Agregar
					</button>
				</form>
		</>
	)
}
