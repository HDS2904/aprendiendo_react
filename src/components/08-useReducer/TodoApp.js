import React, { useReducer } from 'react'
import { todoReducer } from './todoReducer'
import './style.css'

const initialState = [{
	id: new Date().getTime(),
	desc: 'Aprendiendo React',
	done: false
}]

export const TodoApp = () => {
	const [ todos, dispatch ] = useReducer(todoReducer, initialState)
	console.log(todos);
	
	const handleSubmit = (e) => {
		e.preventDefault();

		const newTodo = {
			id: new Date().getTime(),
			desc: 'Nueva actividad',
			done: false
		}

		const action = {
			type: 'add',
			payload: newTodo
		}

		dispatch(action);
	}

	return (
		<div>
			<h1>TodoApp ({todos.length}) </h1>
			<hr />
			<div className="row">
				<div className="col-7">
					<ul className='list-group list-group-flush'> 
						{
							todos.map((todo, index) => (
								<li
									className='list-group-item'
									key={todo.id}
								>
									<p> { index + 1 }. { todo.desc } </p>
									<button
										className='btn btn-danger'
									>
										Borrar
									</button>
								</li>
							))
						}
					</ul>
				</div>
				<div className="col-5">
					<h4>Agregar TODO</h4>
					<hr />
						<form onSubmit={handleSubmit}>
							<input 
								className='form-control'
								type="text" 
								name="description"
								autoComplete='false'
								placeholder='Actividad...'
							/>
							<button
								className='btn btn-outline-primary mt-1 btn-block'
								type='submit'
							>
								Agregar
							</button>
						</form>
				</div>
			</div>
		</div>
	)
}
