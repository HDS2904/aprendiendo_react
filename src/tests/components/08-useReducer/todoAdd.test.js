import { fireEvent, prettyDOM, render, screen } from "@testing-library/react"
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd"

describe('Prueba en <TodoAdd />', () => {

	const handleAddTodo = jest.fn();

	beforeEach( ()=> {
		handleAddTodo.mockClear()
		render(<TodoAdd 
			handleAdd={handleAddTodo}
		/>)
	})
	
	test('Deberia de mostrarse correctamente', () => {

		const wrapper = render( <TodoAdd handleAdd={handleAddTodo} /> )
		expect( wrapper.container ).toMatchSnapshot()

	})

	test('No debe de llamar el handleAddTodo', () => {

		const fun = screen.getByText('Agregar')
		fireEvent.submit(fun)
		expect( handleAddTodo ).toHaveBeenCalledTimes(0)
		
	})

	test('Debe de llamar el handleAddTodo', () => {
		const value = 'Aprendiendo React';
		const input = screen.getByPlaceholderText(/Actividad/i);
		fireEvent.change(input,{
			target: {
				value,
				name: 'description'
			}
		})
		const fun = screen.getByText('Agregar');
		fireEvent.submit(fun);
		expect( handleAddTodo ).toHaveBeenCalledTimes(1)
		expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) )
		expect( handleAddTodo ).toHaveBeenCalledWith({
			id: expect.any(Number),
			desc: value,
			done: false
		})

		expect( input.value ).toBe( '' )
		
	})
	
})