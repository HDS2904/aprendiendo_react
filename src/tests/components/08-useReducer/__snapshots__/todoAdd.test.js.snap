// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <TodoAdd /> Deberia de mostrarse correctamente 1`] = `
<div>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form>
    <input
      autocomplete="off"
      class="form-control"
      name="description"
      placeholder="Actividad..."
      type="text"
      value=""
    />
    <button
      class="btn btn-outline-primary mt-3 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</div>
`;
